service/history/taskPriorityAssigner.go

```diff
@@ -0,0 +1,165 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package history
+
+import (
+	"sync"
+
+	workflow "github.com/uber/cadence/.gen/go/shared"
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/cache"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/metrics"
+	"github.com/uber/cadence/common/quotas"
+)
+
+type (
+	taskPriorityAssigner interface {
+		Assign(queueTask) error
+	}
+
+	taskPriorityAssignerImpl struct {
+		currentClusterName string
+		domainCache        cache.DomainCache
+		config             *Config
+		logger             log.Logger
+		scope              metrics.Scope
+
+		sync.RWMutex
+		rateLimiters map[string]quotas.Limiter
+	}
+)
+
+var _ taskPriorityAssigner = (*taskPriorityAssignerImpl)(nil)
+
+const (
+	taskPriorityActiveTransfer int = iota
+	taskPriorityActiveTimer
+	taskPriorityActiveThrottled
+	taskPriorityReplicationStandby
+)
+
+func newTaskPriorityAssigner(
+	currentClusterName string,
+	domainCache cache.DomainCache,
+	logger log.Logger,
+	metricClient metrics.Client,
+	config *Config,
+) *taskPriorityAssignerImpl {
+	return &taskPriorityAssignerImpl{
+		currentClusterName: currentClusterName,
+		domainCache:        domainCache,
+		config:             config,
+		logger:             logger,
+		scope:              metricClient.Scope(metrics.TaskPriorityAssignerScope),
+		rateLimiters:       make(map[string]quotas.Limiter),
+	}
+}
+
+func (a *taskPriorityAssignerImpl) Assign(
+	task queueTask,
+) error {
+	if task.GetQueueType() == common.ReplicationQueueType {
+		task.SetPriority(taskPriorityReplicationStandby)
+		return nil
+	}
+
+	// timer of transfer task, first check if domain is active or not
+	domainName, active, err := a.getDomainInfo(task.GetDomainID())
+	if err != nil {
+		return err
+	}
+
+	if !active {
+		task.SetPriority(taskPriorityReplicationStandby)
+		return nil
+	}
+
+	if !a.getRateLimiter(domainName).Allow() {
```

Comment:
a bit confused. Why does the priority assigner also do throttling ? Doesn't look like this should belong here

Reply:
It's not doing throttling. This is a just non-blocking call for checking if the number of tasks from a certain domain exceeds limit. If `Allow()` returns false, then it means that domain has generated too many tasks, and laters tasks will be assigned a low priority (`taskPriorityActiveThrottled`). Tasks will still be sent for processing though.

Reply:
This is fine for now. But ideally, rather than using a fixed throttle per domain, you should consider allocating ratios within the available total quota i.e. something like - no single domain can consume more than 10% of available task_processing_capacity on a host (or) lets throttle domainFoo to consume only 1% total available processing capacity on this host.